### ISSUE TITLE
Package usage license clarifications

### ISSUE THREAD

**Issue by canesin** [libMesh/libmesh#224]
Dear developers, first congrats on this great project. Second, would be great if you can clarify the LGPL application in libmesh, can I develop a FEM code using it as shown in the examples and release it as a closed source commercial software ? Which dependencies of libmesh don't allow this, what is the license covering the "standard environment" ?

I would appreciate if these questions are answered, maybe adding it to the repository readme is a good practice.

Regards,
Fábio

**Comment by StefanBruens** [comment]
Unfortunately, concerns regarding METIS/ParMETIS have not been addressed yet:

- Metis is APL 2.0
- libMesh is LGPL-2.1

Both can be included in a package licensed as GPL-3.0, but a libMesh built against Metis is definitely no longer LGPL-2.1.

**Comment by dknez** [comment]
Note also that as we have discussed elsewhere ParMETIS has a restrictive license:
http://glaros.dtc.umn.edu/gkhome/metis/parmetis/download

**Comment by StefanBruens** [comment]
@dknez: correct, ParMETIS has a "non-commercial by default" restriction

**Comment by roystgnr** [comment]
Oh, hell.  I noticed when METIS got the new license, but I didn't realize that ParMETIS hadn't gotten the same treatment.  We'll need to remove ParMETIS from --enable-strict-lgpl builds.

As for METIS itself... we're still LGPL-3.0 compatible there, aren't we, not just GPL-3.0?  LGPL-2.1 isn't compatible with the patent clauses in APL-2.0, but is saved by its own "or any later version" clause.

**Comment by StefanBruens** [comment]
No, libMesh is LGPL-2.1, not LGPL-2.1+, so a combined libMesh is GPL-3.0 (smallest common denominator). See https://www.gnu.org/licenses/gpl-faq.html#AllCompatibility

**Comment by StefanBruens** [comment]
Instead of (Par)METIS, probably scotch/pt-scotch should be bundled. (pt-)scotch is under the CECILL-C license, which is, as far as I understand but AINAL, LGPL-like, thus likely had no influence on a combined work.

**Comment by jwpeterson** [comment]
Regarding CeCILL-C and LGPL compatibility, I found the following [document](https://joinup.ec.europa.eu/sites/default/files/document/2011-12/FAQ-LL-V131-EN.pdf). 

There is a graphic on pg. 16 which shows the compatibility relationships between some of the more popular open source licenses. According to the description of the graphic, "licenses placed on the same horizontal line could cohabit" (LGPL and CeCILL-B, not C, are on the same line).  Furthermore it labels LGPL as "Weak copyleft" while CeCILL-C is "No copyleft".  

My (non-legal) interpretation of this is that attempting license a combined work with LGPL and CeCILL-C licensed components under the LGPL would be invalid, as it would place additional restrictions on the CeCILL-C licensed component. But I would also like to note that I have no idea what I'm talking about...

**Comment by jwpeterson** [comment]
On the other hand, the CeCILL [FAQ](http://www.cecill.info/faq.en.html) itself states:
  
    Q: What are the differences between CeCILL, CeCILL-B and CeCILL-C?
    A: "If you distribute a program that is a modification of a program
    under one of the licenses in the CeCILL family, or that incorporates
    such a program, totally or partially, [and that CeCILL-licensed]
    program is under CeCILL-C, you can also distribute your program
    under any license but you have to distribute under CeCILL the
    possible modifications you did to the original source code (and
    mention the use you made of the initial program)."

The "any license" language seems overly broad to me (?) but it seems to suggest we would be allowed to license libmesh-with-contrib-ptscotch under the LGPL.

And, specifically to our case:

    Q: I only modified only the Makefile of a software under CeCILL-C,
    or I did nothing but rename one function. Must I really redistribute
    that?
    A: The spirit of the licence is that the modifications that should
    be given back are the ones that could be useful for the
    community. The text of CeCILL-C indicates "such that such
    modifications may be carried over in a subsequent version of the
    Software".

Looking at their build system, it's likely that will have to make some changes in order to automatet it, but it's unlikely that they would be useful to the community at large.

**Comment by roystgnr** [comment]
> it seems to suggest we would be allowed to license libmesh-with-contrib-ptscotch under the LGPL.

Not quite, I think.  We would be able to license libmesh.so under the LGPL, while simultaneously distributing a compatibly-compiled libptscotch.so under CeCILL-C, but we wouldn't be able to relicense PT-Scotch as LGPL.  It's basically the same situation as when you distribute a commercial code linking to an LGPL library, only here there's nothing commercial, just two mutually incompatible copyleft codes.

**Comment by roystgnr** [comment]
This isn't entirely a new situation.  If you configure libMesh with --disable-strict-lgpl then you get other libfoo.so that can be linked to our LGPL code but that can only be redistributed if you also meet their peculiar licensing requirements.  We ought to nearly be in the same boat even with METIS (except for GPLv3/GPLv3+/AGPLv3 exceptions) according to @StefanBruens, who I believe to be correct.

What's unique about scotch is that it's license is *nearly* as good as the LGPL, so we'd love to be able to enable it even with --enable-strict-lgpl, but it's *not* LGPL relicensable (which makes a difference, not just because multi-license redistribution is more confusing, but because CeCILL-C isn't GPL compatible, and users of --enable-strict-lgpl might reasonably expect to build GPL codes with the result).

**Comment by roystgnr** [comment]
Reopening until we get the ParMETIS/--enable-strict-lgpl issue straight.

Looking at scotch, it seems to be an orphaned project?  The team and project web pages are dead, and the main page is advertising their latest release as of 5 years ago.  The [forge page](https://gforge.inria.fr/projects/scotch/) is still active, and shows a few bugfix releases in 2015 and 2016, but nothing major.  Their tracker includes 5 patch submissions in 2017, none of which has received any attention.

I harassed @bangerth, and his suggestion was [Zoltan](http://www.cs.sandia.gov/Zoltan/ug_html/ug.html).  They're also not getting a lot of recent development, but they do have the basics (parallel, 64-bit) we need, and as long as they're part of Trilinos they shouldn't be subject to bitrot at least.  Trilinos inclusion also means we'd have less trouble hooking them into our build system, they're BSD licensed, and they have a shim to Scotch so we could probably use them to experiment with Scotch "under the hood" after only writing one Partitioner subclass.
